File: code/csharp/ExceptionHandling.cs
// Throwing, catching, filtering, and wrapping exceptions.

using System;
using System.Collections.Generic;
using System.IO;

class ValidationException : Exception
{
    public string Field { get; }

    public ValidationException(string field, string message) : base(message) => Field = field;

    public ValidationException(string field, string message, Exception inner)
        : base(message, inner) => Field = field;
}

class ExceptionHandling
{
    static int ParseZone(string raw)
    {
        try
        {
            int zone = int.Parse(raw);
            if (zone is < 1 or > 6)
            {
                throw new ValidationException(nameof(zone), $"zone {zone} is outside 1-6");
            }
            return zone;
        }
        catch (FormatException error)
        {
            // Wrap the low-level failure in something the caller can act on.
            throw new ValidationException(nameof(zone), $"'{raw}' is not a number", error);
        }
    }

    static void Main()
    {
        foreach (string raw in new[] { "3", "9", "east", "5" })
        {
            try
            {
                Console.WriteLine($"{raw,-5} -> zone {ParseZone(raw)}");
            }
            catch (ValidationException error) when (error.InnerException is not null)
            {
                Console.WriteLine($"{raw,-5} -> {error.Message} (from {error.InnerException.GetType().Name})");
            }
            catch (ValidationException error)
            {
                Console.WriteLine($"{raw,-5} -> {error.Message}");
            }
        }

        // finally runs whether or not the try block succeeded.
        var opened = new List<string>();
        try
        {
            opened.Add("report.csv");
            throw new IOException("the disk went away");
        }
        catch (IOException error)
        {
            Console.WriteLine($"caught {error.Message}");
        }
        finally
        {
            opened.Clear();
            Console.WriteLine($"cleaned up, {opened.Count} handles left open");
        }
    }
}

File: code/dart/exception_handling.dart
// Throwing, catching by type, rethrowing, and cleaning up in finally.

class ValidationException implements Exception {
  ValidationException(this.field, this.message, [this.cause]);

  final String field;
  final String message;
  final Object? cause;

  @override
  String toString() => 'ValidationException($field): $message';
}

int parseZone(String raw) {
  try {
    final zone = int.parse(raw);
    if (zone < 1 || zone > 6) {
      throw ValidationException('zone', 'zone $zone is outside 1-6');
    }
    return zone;
  } on FormatException catch (error) {
    // Wrap the low-level failure in something the caller can act on.
    throw ValidationException('zone', '"$raw" is not a number', error);
  }
}

Future<String> loadWithRetry(int attempts) async {
  for (var attempt = 1; attempt <= attempts; attempt++) {
    try {
      if (attempt < attempts) {
        throw StateError('attempt $attempt failed');
      }
      return 'loaded on attempt $attempt';
    } on StateError catch (error) {
      print('  ${error.message}');
      if (attempt == attempts) {
        rethrow; // out of retries: let the caller deal with it
      }
    }
  }
  throw StateError('unreachable');
}

Future<void> main() async {
  for (final raw in ['3', '9', 'east']) {
    try {
      print('${raw.padRight(6)} -> zone ${parseZone(raw)}');
    } on ValidationException catch (error) {
      final because = error.cause == null ? '' : ' (from ${error.cause.runtimeType})';
      print('${raw.padRight(6)} -> ${error.message}$because');
    }
  }

  print(await loadWithRetry(3));

  final open = <String>['report.csv'];
  try {
    throw const FormatException('the file ended halfway through a row');
  } on FormatException catch (error, stack) {
    print('caught: ${error.message}');
    print('stack has ${stack.toString().split('\n').length} frames');
  } finally {
    open.clear();
    print('cleaned up, ${open.length} handles left open');
  }
}
